File: src/views/project/project_list_task.py
import gi
from gi.repository import Gtk, Adw, GLib, Gdk
from time import sleep
from datetime import datetime
from threading import Thread

from iplan.db.models.task import Task
from iplan.db.operations.task import update_task, delete_task


@Gtk.Template(resource_path="/ir/imansalmani/iplan/ui/project/project_list_task.ui")
class ProjectListTask(Gtk.ListBoxRow):
    __gtype_name__ = "ProjectListTask"
    task: Task
    checkbox: Gtk.CheckButton = Gtk.Template.Child()
    name_entry: Gtk.Entry = Gtk.Template.Child()
    name_button: Gtk.Button = Gtk.Template.Child()
    timer_toggle_button: Gtk.Button = Gtk.Template.Child()
    timer_button_content: Adw.ButtonContent = Gtk.Template.Child()
    timer_value = 0
    moving_out = False  # When drag

    def __init__(self, task):
        super().__init__()
        self.task = task

        self.checkbox.set_active(self.task.done)
        self.name_button.get_child().set_text(self.task.name)
        self.name_button.set_tooltip_text(self.task.name)
        self.name_entry.get_buffer().set_text(self.task.name, -1)
        name_entry_controller = Gtk.EventControllerKey()
        name_entry_controller.connect(
            "key-released", self.name_entry_controller_key_released_cb
        )
        self.name_entry.add_controller(name_entry_controller)

        duration = task.get_duration()
        if duration:
            self.timer_button_content.set_label(task.duration_to_text(duration))

        if task.done:
            self.timer_toggle_button.set_sensitive(False)
        else:
            self.timer_toggle_button.connect(
                "toggled", self.timer_toggle_button_toggled_cb
            )

            # starting timer if last duration not finished
            last_time = task.get_last_time()
            if last_time:
                if not last_time[1]:
                    start = datetime.fromtimestamp(last_time[0])
                    self.timer_value = datetime.now() - start
                    self.timer_toggle_button.set_active(True)

    # Name
    @Gtk.Template.Callback()
    def name_button_clicked_cb(self, *args):
        self.name_button.set_visible(False)  # Entry visible param binded to this
        self.name_entry.grab_focus_without_selecting()

    @Gtk.Template.Callback()
    def name_entry_activate_cb(self, *args):
        self.name_button.set_visible(True)  # Entry visible param binded to this
        self.task.name = self.name_entry.get_buffer().get_text()
        self.name_button.get_child().set_text(self.task.name)
        self.name_button.set_tooltip_text(self.task.name)
        update_task(self.task)

    @Gtk.Template.Callback()
    def name_entry_icon_press_cb(self, *args):  # Cancel name editing
        self.name_button.set_visible(not self.name_button.get_visible())
        self.name_entry.get_buffer().set_text(self.task.name, -1)

    def name_entry_controller_key_released_cb(self, controller, keyval, keycode, state):
        if keycode == 9:  # Escape
            self.name_entry.emit("icon-press", Gtk.EntryIconPosition.SECONDARY)

    # Delete
    @Gtk.Template.Callback()
    def delete_button_clicked_cb(self, *args):
        window = self.get_root()
        toast_name = self.task.name
        if len(toast_name) > 10:
            toast_name = f"{toast_name[0:9]}..."
        toast = Adw.Toast.new(f'"{toast_name}" deleted')
        toast.set_button_label("Undo")
        toast.connect("button-clicked", self.delete_toast_button_clicked_cb)
        toast.connect("dismissed", self.delete_toast_dismissed_cb)
        window.toast_overlay.add_toast(toast)
        self.task.suspended = True
        update_task(self.task)
        # Prevent from scroll up after suspend row
        upper_task = self.get_parent().get_row_at_index(self.get_index() - 1)
        if upper_task:
            window.set_focus(upper_task)
        else:
            window.set_focus(self.get_parent().get_row_at_index(1))
        self.changed()

    def delete_toast_button_clicked_cb(self, *args):  # Undo button
        self.task.suspended = False
        update_task(self.task)

        window = self.get_root()
        if window:  # This happens after open another project
            self.changed()
            window.set_focus(self)

    def delete_toast_dismissed_cb(self, *args):
        if not self.task.suspended:  # Checking Undo button
            return

        delete_task(self.task)
        tasks_box = self.get_parent()
        # TODO: tasks_box should removed after open another project
        # this should have None check after fixing memory leak
        # Decrease upper tasks position
        for i in range(0, self.get_index()):
            tasks_box.get_row_at_index(i).task.position -= 1
        # Prevent from scroll top in vertical mode
        upper_task = tasks_box.get_row_at_index(self.get_index() - 1)
        if upper_task:
            self.get_root().set_focus(upper_task)
        tasks_box.remove(self)

    # Open
    # def open_task(self):
    #     window = self.get_root()
    #     modal = TaskModal(self.task)
    #     modal.set_transient_for(window)
    #     modal.present()

    # Done
    @Gtk.Template.Callback()
    def done_check_button_toggled_cb(self, sender):
        active = sender.get_active()

        if self.task.done == active:  # This happens in fetch done tasks
            return

        self.task.done = active
        update_task(self.task)

        if active:
            # Stop timer and disconnect handler
            if self.timer_toggle_button.get_active():
                self.timer_toggle_button.set_active(False)
            self.timer_toggle_button.disconnect_by_func(
                self.timer_toggle_button_toggled_cb
            )

            # Remove or filter row
            self.activate_action("task.done", GLib.Variant("i", self.get_index()))
        else:
            self.timer_toggle_button.set_sensitive(True)
            self.timer_toggle_button.connect(
                "toggled", self.timer_toggle_button_toggled_cb
            )

    # Timer
    def timer_toggle_button_toggled_cb(self, *args):
        """
        - Start timer if is active.
            timer daemon checking timer toggle button
            active property every second and will stop if is False.
        - Save duration if is not active
        """
        if self.timer_toggle_button.get_active():
            self.timer_toggle_button.add_css_class("destructive-action")
            thread = Thread(target=self.start_timer)
            thread.daemon = True
            thread.start()
        else:
            self.timer_toggle_button.remove_css_class("destructive-action")
            self.save_duration()

    def start_timer(self):
        # add duration record if task don't have unfinished timer
        if self.timer_value:
            start = datetime.fromtimestamp(self.task.get_last_time()[0])
        else:
            start = datetime.now()
            self.task.duration += f"{start.timestamp()},0;"
            update_task(self.task)

        while self.timer_toggle_button.get_active():
            now = datetime.now()
            self.timer_value = now - start
            GLib.idle_add(
                lambda: self.timer_button_content.set_label(
                    self.task.duration_to_text(self.timer_value.seconds)
                )
            )
            sleep(0.1)

    def save_duration(self):
        self.task.duration = (
            self.task.duration[0:-2] + str(self.timer_value.seconds) + ";"
        )
        self.timer_button_content.set_label(self.task.get_duration_text())
        self.timer_value = 0
        update_task(self.task)
        self.activate_action("project.update")

    # Drag
    @Gtk.Template.Callback()
    def drag_prepare_cb(
        self, drag_source: Gtk.DragSource, x: float, y: float
    ) -> Gdk.ContentProvider:
        if not self.name_entry.get_visible():
            return Gdk.ContentProvider.new_for_value(self)

    @Gtk.Template.Callback()
    def drag_begin_cb(self, drag_source: Gtk.DragSource, drag: Gdk.Drag):
        # allocation = self.get_allocation()
        # drag_widget = Gtk.ListBox()
        # drag_widget.set_size_request(240, allocation.height)

        # drag_row = ProjectListTask(self.task)
        # drag_row.delete_button.set_visible(False)
        # drag_row.timer.set_visible(False)
        # drag_widget.append(drag_row)
        # drag_widget.drag_highlight_row(drag_row)
        # drag_row.set_size_request(240, 64)
        self.get_parent().drag_highlight_row(self)
        drag_icon = Gtk.DragIcon.get_for_drag(drag)
        drag_icon.props.child = Gtk.Label()
        drag.set_hotspot(0, 0)

    @Gtk.Template.Callback()
    def drag_cancel_cb(self, drag_source: Gtk.DragSource, drag: Gdk.Drag, reason):
        self.moving_out = False
        self.changed()
        return False

    @Gtk.Template.Callback()
    def drag_end_cb(self, drag_source: Gtk.DragSource, drag: Gdk.Drag, delete_data):
        if self.get_root():
            self.get_root().sidebar.projects_section.select_active_project()
        return False


class TaskModal(Adw.Window):
    __gtype_name__ = "TaskModal"

    def __init__(self, task: Task):
        super().__init__()
        self.set_modal(True)
        self.set_size_request(480, 480)

        content = Gtk.Box()
        content.set_orientation(Gtk.Orientation.VERTICAL)
        self.set_content(content)

        header = Adw.HeaderBar()
        header.add_css_class("flat")
        header.set_title_widget(Gtk.Label())
        content.append(header)

        title = Gtk.Label.new(task.name)
        title.add_css_class("title-1")
        content.append(title)

File: src/views/project/project_list_task.rs
use gtk::{gdk, glib, glib::once_cell::sync::Lazy, prelude::*, subclass::prelude::*};
use std::cell::{Cell, RefCell};
use std::thread;
use std::time::Duration;

use crate::db::models::{Record, Task};
use crate::db::operations::{create_record, delete_task, read_records, update_record, update_task};
use crate::views::IPlanWindow;

mod imp {
    use super::*;

    #[derive(Default, gtk::CompositeTemplate)]
    #[template(resource = "/ir/imansalmani/iplan/ui/project/project_list_task.ui")]
    pub struct ProjectListTask {
        pub task: RefCell<Task>,
        pub moving_out: Cell<bool>,
        #[template_child]
        pub checkbox: TemplateChild<gtk::CheckButton>,
        #[template_child]
        pub name_button: TemplateChild<gtk::Button>,
        #[template_child]
        pub name_entry: TemplateChild<gtk::Entry>,
        #[template_child]
        pub timer_toggle_button: TemplateChild<gtk::ToggleButton>,
        pub timer_toggle_button_handler_id: RefCell<Option<gtk::glib::SignalHandlerId>>,
        #[template_child]
        pub timer_button_content: TemplateChild<adw::ButtonContent>,
    }

    #[glib::object_subclass]
    impl ObjectSubclass for ProjectListTask {
        const NAME: &'static str = "ProjectListTask";
        type Type = super::ProjectListTask;
        type ParentType = gtk::ListBoxRow;

        fn class_init(klass: &mut Self::Class) {
            klass.bind_template();
            klass.bind_template_instance_callbacks();
        }

        fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
            obj.init_template();
        }
    }

    impl ObjectImpl for ProjectListTask {
        fn properties() -> &'static [glib::ParamSpec] {
            static PROPERTIES: Lazy<Vec<glib::ParamSpec>> =
                Lazy::new(|| vec![glib::ParamSpecObject::builder::<Task>("task").build()]);
            PROPERTIES.as_ref()
        }

        fn set_property(&self, _id: usize, value: &glib::Value, pspec: &glib::ParamSpec) {
            match pspec.name() {
                "task" => {
                    let value = value.get::<Task>().expect("value must be a Task");
                    self.task.replace(value);
                }
                _ => unimplemented!(),
            }
        }

        fn property(&self, _id: usize, pspec: &glib::ParamSpec) -> glib::Value {
            match pspec.name() {
                "task" => self.task.borrow().to_value(),
                _ => unimplemented!(),
            }
        }
    }
    impl WidgetImpl for ProjectListTask {}
    impl ListBoxRowImpl for ProjectListTask {}
}

glib::wrapper! {
    pub struct ProjectListTask(ObjectSubclass<imp::ProjectListTask>)
        @extends gtk::Widget, gtk::ListBoxRow,
        @implements gtk::Buildable;
}

#[gtk::template_callbacks]
impl ProjectListTask {
    pub fn new(task: Task) -> Self {
        glib::Object::builder().property("task", task).build()
    }

    pub fn init_widgets(&self) {
        let imp = self.imp();
        let task = self.task();
        let task_name = task.name();

        imp.checkbox.set_active(task.done());
        imp.name_button
            .child()
            .unwrap()
            .downcast::<gtk::Label>()
            .unwrap()
            .set_text(&task_name);
        imp.name_entry.buffer().set_text(&task_name);
        let name_entry_controller = gtk::EventControllerKey::new();
        name_entry_controller.connect_key_released(glib::clone!(
        @weak self as obj =>
        move |_controller, _keyval, keycode, _state| {
            if keycode == 9 {   // Escape
                let imp = obj.imp();
                imp.name_button.set_visible(true);
                imp.name_entry.buffer().set_text(&obj.task().name());
            }
        }));
        imp.name_entry.add_controller(&name_entry_controller);
        if let Some(duration) = task.duration() {
            imp.timer_button_content
                .set_label(&Record::duration_display(duration));
        }
        if task.done() {
            imp.timer_toggle_button.set_sensitive(false);
        } else {
            let handler_id = imp.timer_toggle_button.connect_toggled(glib::clone!(
                @weak self as obj =>
                move |button| obj.handle_timer_toggle_button_toggled(button)));
            imp.timer_toggle_button_handler_id.replace(Some(handler_id));
            // Starting timer if have not finished record
            let records =
                read_records(task.id(), true, None, None).expect("Faield to read records");
            if !records.is_empty() {
                imp.timer_toggle_button.set_active(true)
            }
        }
    }

    pub fn task(&self) -> Task {
        self.property("task")
    }

    #[template_callback]
    fn handle_done_check_button_toggled(&self, button: gtk::CheckButton) {
        let imp = self.imp();
        let task = self.task();
        let is_active = button.is_active();

        if task.done() != is_active {
            // This happens in fetch done tasks
            task.set_property("done", is_active);
            self.set_property("task", task);
            update_task(self.task()).expect("Failed to update task");

            if is_active {
                imp.timer_toggle_button.set_active(false);
                imp.timer_toggle_button.set_sensitive(false);

                self.activate_action("task.done", Some(&self.index().to_variant()))
                    .expect("Failed to activate task.done action");
            } else {
                imp.timer_toggle_button.set_sensitive(true);
                if imp.timer_toggle_button_handler_id.borrow().is_none() {
                    let handler_id = imp.timer_toggle_button.connect_toggled(glib::clone!(
                        @weak self as obj =>
                        move |button| obj.handle_timer_toggle_button_toggled(button)));
                    imp.timer_toggle_button_handler_id.replace(Some(handler_id));
                }
            }
        }
    }

    #[template_callback]
    fn handle_name_button_clicked(&self, button: gtk::Button) {
        button.set_visible(false); // Entry visible param binded to this
        self.imp().name_entry.grab_focus_without_selecting();
    }

    #[template_callback]
    fn handle_name_entry_activate(&self, entry: gtk::Entry) {
        let name = entry.buffer().text();
        let task = self.task();
        let imp = self.imp();
        imp.name_button
            .child()
            .unwrap()
            .downcast::<gtk::Label>()
            .unwrap()
            .set_text(&name);
        imp.name_button.set_visible(true);
        imp.name_button.set_tooltip_text(Some(&name));
        task.set_property("name", name);
        update_task(task).expect("Failed to update task");
    }

    #[template_callback]
    fn handle_name_entry_icon_press(&self, _: gtk::EntryIconPosition) {
        let imp = self.imp();
        imp.name_button.set_visible(true);
        imp.name_entry.buffer().set_text(&self.task().name());
    }

    fn handle_timer_toggle_button_toggled(&self, button: &gtk::ToggleButton) {
        let task = self.task();
        let records = read_records(task.id(), true, None, None).expect("Faield to read records");
        let record = if records.is_empty() {
            create_record(glib::DateTime::now_local().unwrap().to_unix(), task.id())
                .expect("Failed to create record")
        } else {
            let record = records.get(0).unwrap().clone();
            record.set_property(
                "duration",
                glib::DateTime::now_local().unwrap().to_unix() - record.start(),
            );
            record
        };

        if button.is_active() {
            button.add_css_class("destructive-action");

            let (tx, rx) = glib::MainContext::channel(glib::PRIORITY_DEFAULT);
            let mut duration = record.duration();
            thread::spawn(move || loop {
                if let Err(_) = tx.send(duration.to_string()) {
                    break;
                }
                thread::sleep(Duration::from_secs(1));
                duration += 1;
            });
            rx.attach(
                None,
                glib::clone!(
                    @weak button => @default-return glib::Continue(false),
                    move |text| {
                        if button.is_active() {
                            let button_content = button.child()
                                .and_downcast::<adw::ButtonContent>()
                                .unwrap();
                            button_content.set_label(
                                &Record::duration_display(text.parse::<i64>().unwrap())
                            );
                            glib::Continue(true)
                        } else {
                            glib::Continue(false)
                        }
                    }
                ),
            );
        } else {
            button.remove_css_class("destructive-action");
            record.set_property(
                "duration",
                glib::DateTime::now_local().unwrap().to_unix() - record.start(),
            );
            update_record(&record).expect("Failed to update record");
            self.imp()
                .timer_button_content
                .set_label(&Record::duration_display(
                    task.duration()
                        .expect("Task duration cannot be 0 at this point"),
                ));
            self.activate_action("project.update", None)
                .expect("Failed to send project.update");
        }
    }

    #[template_callback]
    fn handle_delete_button_clicked(&self, _button: gtk::Button) {
        let task = self.task();
        let mut toast_name = task.name();
        toast_name.shrink_to(10);
        let toast = adw::Toast::builder()
            .title(&toast_name)
            .button_label("Undo")
            .build();

        toast.connect_button_clicked(glib::clone!(
        @weak self as obj =>
        move |_toast| {
            let task = obj.task();
            task.set_property("suspended", false);
            update_task(task).expect("Failed to update task");
            if obj.parent().is_some() {
                obj.changed();
                obj.grab_focus();
            }
        }));
        toast.connect_dismissed(glib::clone!(
            @strong self as obj =>
            move |_toast| {
                let task = obj.task();
                if task.suspended() {    // Checking Undo button
                    delete_task(task.id(), task.list(), task.position())
                        .expect("Failed to delete task");
                    if let Some(tasks_box) = obj.parent() {    // check for project active changed
                        let tasks_box = tasks_box.downcast::<gtk::ListBox>().unwrap();
                        for i in 0..obj.index() {
                            let row = tasks_box.row_at_index(i).unwrap();
                            let row_task: Task = row.property("task");
                            row_task.set_property("position", row_task.position() - 1);
                        }
                        let upper_row = obj.parent()
                            .and_downcast::<gtk::ListBox>()
                            .unwrap()
                            .row_at_index(obj.index() - 1);
                        if let Some(upper_row) = upper_row {upper_row.grab_focus();}
                        tasks_box.remove(&obj);
                    }
                }
            }
        ));
        let window = self.root().and_downcast::<IPlanWindow>().unwrap();
        window.imp().toast_overlay.add_toast(&toast);
        task.set_property("suspended", true);
        self.set_property("task", &task);
        update_task(task).expect("Failed to update task");
        let upper_row = self
            .parent()
            .and_downcast::<gtk::ListBox>()
            .unwrap()
            .row_at_index(self.index() - 1);
        if let Some(upper_row) = upper_row {
            upper_row.grab_focus();
        }
        self.changed();
    }

    #[template_callback]
    fn handle_drag_prepare(&self, _x: f64, _y: f64) -> Option<gdk::ContentProvider> {
        let name_entry = self.imp().name_entry.get();
        if WidgetExt::is_visible(&name_entry) {
            None
        } else {
            Some(gdk::ContentProvider::for_value(&self.to_value()))
        }
    }

    #[template_callback]
    fn handle_drag_begin(&self, drag: gdk::Drag) {
        self.parent()
            .and_downcast::<gtk::ListBox>()
            .unwrap()
            .drag_highlight_row(self);
        let drag_icon: gtk::DragIcon = gtk::DragIcon::for_drag(&drag).downcast().unwrap();
        let label = gtk::Label::builder().label("").build();
        drag_icon.set_child(Some(&label));
        drag.set_hotspot(0, 0);
    }

    #[template_callback]
    fn handle_drag_cancel(&self, _drag: gdk::Drag) -> bool {
        self.imp().moving_out.set(false);
        self.changed();
        false
    }
}
